packages/next/src/build/babel/preset.ts

```diff
@@ -57,12 +75,112 @@ function supportsStaticESM(caller: any): boolean {
   return !!caller?.supportsStaticESM
 }
 
+/**
+ * HACK: A drop-in replacement for `@babel/preset-typescript` that only enables
+ * `@babel/plugin-syntax-typescript` and does not transform the typescript.
+ *
+ * This is used for standalone mode, where Babel is being used alongside SWC
+ * (i.e. Turbopack or with React Compiler on webpack).
+ *
+ * This should match the logic/behavior here:
+ * https://github.com/babel/babel/blob/7f57d3a2e97b7e2800fb82cff9284a3591377971/packages/babel-preset-typescript/src/index.ts#L63
+ */
+function presetTypescriptSyntaxOnly(_api: unknown, options: any) {
+  const { allExtensions, ignoreExtensions, ...restOptions } = options
+  const disableExtensionDetect = allExtensions || ignoreExtensions
+
+  function getPlugins(isTSX: boolean, disallowAmbiguousJSXLike: boolean) {
+    return [
+      [
+        require('next/dist/compiled/babel/plugin-syntax-typescript') as typeof import('next/dist/compiled/babel/plugin-syntax-typescript'),
+        {
+          isTSX,
+          disallowAmbiguousJSXLike,
+          ...restOptions,
+        },
+      ],
+    ]
+  }
+
+  return {
+    plugins: [],
+    overrides: disableExtensionDetect
+      ? [
+          {
+            plugins: getPlugins(
+              options.isTSX,
+              options.disallowAmbiguousJSXLike
+            ),
+          },
+        ]
+      : // Only set 'test' if explicitly requested, since it requires that
+        // Babel is being called with a filename.
+        [
+          {
+            test: /\.ts$/,
+            plugins: getPlugins(false, false),
+          },
+          {
+            test: /\.mts$/,
+            sourceType: 'module',
+            plugins: getPlugins(false, true),
+          },
+          {
+            test: /\.cts$/,
+            sourceType: 'unambiguous',
+            plugins: getPlugins(false, true),
+          },
+          {
+            test: /\.tsx$/,
+            plugins: getPlugins(true, false),
+          },
+        ],
+  }
+}
+
 export default (
   api: any,
   options: NextBabelPresetOptions = {}
 ): BabelPreset => {
-  const supportsESM = api.caller(supportsStaticESM)
+  const isStandalone = api.caller(
+    // NOTE: `transformMode` may be undefined if the user configured `babel-loader` themselves. In
+    // this case, we should assume we're in 'default' mode.
+    (caller: any) => caller.transformMode === 'standalone'
```

Comment:
I think you need to do this like below:

```suggestion
    (caller: any) => !!caller && caller.transformMode === 'standalone'
```

I got an error when upgrading to the latest canary related to this. Changing that line fixed it. It's not from running the Next.js app itself, but I have a localization tool that's also using Babel and I'm reusing the `next/babel` preset.

```
Cannot read properties of undefined (reading 'transformMode') (While processing: "/home/runner/work/my-project/node_modules/next/babel.js")
```

Reply:
Added a tiny PR for this in https://github.com/vercel/next.js/pull/84154.